webapp/src/components/react_select_setting.jsx

```diff
@@ -0,0 +1,157 @@
+
+export default class ReactSelectSetting extends React.PureComponent {
+    static propTypes = {
+        name: PropTypes.string,
+        onChange: PropTypes.func,
+        theme: PropTypes.object.isRequired,
+        isClearable: PropTypes.bool,
+        options: PropTypes.array.isRequired,
+        value: PropTypes.oneOfType([
+            PropTypes.object,
+            PropTypes.array,
+            PropTypes.string,
+        ]),
+        addValidate: PropTypes.func,
+        removeValidate: PropTypes.func,
+        required: PropTypes.bool,
+        allowUserDefinedValue: PropTypes.bool,
+        limitOptions: PropTypes.bool,
+        resetInvalidOnChange: PropTypes.bool,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {invalid: false};
+    }
+
+    componentDidMount() {
+        if (this.props.addValidate && this.props.name) {
+            this.props.addValidate(this.props.name, this.isValid);
+        }
+    }
+
+    componentWillUnmount() {
+        if (this.props.removeValidate && this.props.name) {
+            this.props.removeValidate(this.props.name);
+        }
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (prevState.invalid && (this.props.value && this.props.value.value) !== (prevProps.value && prevProps.value.value)) {
+            this.setState({invalid: false}); //eslint-disable-line react/no-did-update-set-state
+        }
+    }
+
+    handleChange = (value) => {
+        if (this.props.onChange) {
+            if (Array.isArray(value)) {
+                this.props.onChange(this.props.name, value.map((x) => x.value));
+            } else {
+                const newValue = value ? value.value : null;
+                this.props.onChange(this.props.name, newValue);
+            }
+        }
+
+        if (this.props.resetInvalidOnChange) {
+            this.setState({invalid: false});
+        }
+    };
+
+    // Standard search term matching plus reducing to < 100 items
+    filterOptions = (input) => {
+        let options = this.props.options;
+        if (input) {
+            options = options.filter((x) => x.label.toUpperCase().includes(input.toUpperCase()));
+        }
+        return Promise.resolve(options.slice(0, MAX_NUM_OPTIONS));
+    };
+
+    isValid = () => {
+        if (!this.props.required) {
+            return true;
+        }
+
+        const valid = Boolean(this.props.value);
+
+        this.setState({invalid: !valid});
+        return valid;
+    };
+
+    render() {
+        const requiredMsg = 'This field is required.';
+        let validationError = null;
+        if (this.props.required && this.state.invalid) {
+            validationError = (
+                <p className='help-text error-text'>
+                    <span>{requiredMsg}</span>
+                </p>
+            );
+        }
+
+        let selectComponent = null;
+        if (this.props.limitOptions && this.props.options.size > MAX_NUM_OPTIONS) {
```

Comment:
I would say yes, this is a better option, and just set the default in this component if MAX_NUM_OPTIONS is not set by the parent.  This code was stolen from `jira` and I tried minimizing changes to the code.  Hopefully we can compare the two and create a generic shared component because this one is so useful.

Reply:
@larkox It's a sane default to minimize the number of DOM elements present at any given time. Someone had 5000+ Jira projects which led to us making this component work this way :)

The limit only applies to what is currently selectable in the component. If you start typing, options that were not visible at first (e.g. their first letter is fairly late in the alphabet) will show after the results have been filtered.